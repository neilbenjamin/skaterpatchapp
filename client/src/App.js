
//Main app output

//Imprts and components
import React, { useState } from "react";
import { BrowserRouter as Router, Routes, Route } from "react-router-dom";
import UserRegistration from "./RegisterLogin/userRegistration";
import UserLogin from "./RegisterLogin/userLogin";
import AdminDashboard from "./pages/AdminDashboard";
import SkaterDashboard from "./pages/SkaterDashboard";
// Admin Page Routes
import AdminContactPage from "./pages/AdminContactPage";
import AdminUpdateUserPage from "./pages/AdminUpdateUserPage";
import AdminProfilePage from "./pages/AdminProfilePage";
import AdminHomePage from "./pages/AdminHomePage";
//Skater Pages
import SkaterContactPage from "./pages/SkaterContactPage";
import SkaterProfileDisplayPage from "./pages/SkaterProfileDisplayPage";
import SkaterHomePage from "./pages/SkaterHomePage";

//Mainstate of the application.
function App() {
  const [isLoggedIn, setIsLoggedIn] = useState(false);
  const [userRole, setUserRole] = useState(null);

  //Another challenging quandry based on correct JWT Token extraction. "role" is defined in login
  // passed into this function as a JWT from localStorage, containing the userId and role (skater or admin).
  //Rememeber on default, everyone is skater, until a super user (you or me) decides to change the skater's role
  //To a admin. This is currently done inside the application using the router.put('/update-role/:userId', updateRole); route
  //as defined in the server and necessary controllers and schema to update the role default role from skater to
  //admin. These rules are set up in the controllers where the JWT's are primarily initialized.
  const handleLoginSuccess = (role) => {
    setIsLoggedIn(true);
    setUserRole(role);
    localStorage.setItem("userRole", role); // Storing user role in local storage
  };

  const handleLogout = () => {
    setIsLoggedIn(false);
    setUserRole(null);
    localStorage.removeItem("token");
    localStorage.removeItem("userRole");
  };

  return (
    // Main Router and display
    <Router>
      <div>
        {/* Ternary operator used to set dashboard paths based on user role */}
        {!isLoggedIn ? (
          <>
            <UserRegistration />
            <UserLogin onLoginSuccess={handleLoginSuccess} />
          </>
        ) : userRole === "admin" ? (
          <AdminDashboard onLogout={handleLogout} />
        ) : (
          <SkaterDashboard onLogout={handleLogout} />
        )}
        <Routes>
          {/* AdminDashBoard */}
          <Route path="/admin/home" element={<AdminHomePage />} />
          <Route path="/admin/new-user" element={<AdminUpdateUserPage />} />
          <Route path="/admin/profile" element={<AdminProfilePage />} />
          <Route path="/admin/contact" element={<AdminContactPage />} />
          {/* SkaterDashBoard */}
          <Route path="/skater/contact" element={<SkaterContactPage />} />
          <Route
            path="/skater/profile"
            element={<SkaterProfileDisplayPage />}
          />
          <Route path="/skater/home" element={<SkaterHomePage />} />?
        </Routes>
      </div>
    </Router>
  );
}

export default App;
