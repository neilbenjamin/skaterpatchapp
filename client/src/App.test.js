import { render } from "@testing-library/react";
import App from "./App";

test("renders App snapshot", () => {
  //Utiising the asFragmant property in th render Object which takes a snapshot of HTML structure of the component selected. 
  const { asFragment } = render(<App />);
  expect(asFragment()).toMatchSnapshot();
});
