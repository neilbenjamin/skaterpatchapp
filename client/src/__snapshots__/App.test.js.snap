// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders App snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="container my-4"
    >
      <div
        class="row"
      >
        <div
          class="col-md-6 offset-md-3"
        >
          <div
            class="card"
          >
            <div
              class="card-body"
            >
              <h2
                class="card-title mb-4 text-center"
              >
                Register
              </h2>
              <form>
                <div
                  class="mb-3"
                >
                  <label
                    class="form-label"
                    for="registration-name"
                  >
                    Name:
                  </label>
                  <input
                    class="form-control"
                    id="registration-name"
                    placeholder="Enter your name"
                    required=""
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="mb-3"
                >
                  <label
                    class="form-label"
                    for="registration-surname"
                  >
                    Surname:
                  </label>
                  <input
                    class="form-control"
                    id="registration-surname"
                    placeholder="Enter your surname"
                    required=""
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="mb-3"
                >
                  <label
                    class="form-label"
                    for="registration-email"
                  >
                    Email:
                  </label>
                  <input
                    class="form-control"
                    id="registration-email"
                    placeholder="Enter your email"
                    required=""
                    type="email"
                    value=""
                  />
                </div>
                <div
                  class="mb-3"
                >
                  <label
                    class="form-label"
                    for="registration-password"
                  >
                    Password:
                  </label>
                  <input
                    class="form-control"
                    id="registration-password"
                    placeholder="Create a password"
                    required=""
                    type="password"
                    value=""
                  />
                </div>
                <button
                  class="btn btn-primary w-100"
                  type="button"
                >
                  Register
                </button>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="container my-4"
    >
      <div
        class="row"
      >
        <div
          class="col-md-6 offset-md-3"
        >
          <div
            class="card"
          >
            <div
              class="card-body"
            >
              <h2
                class="card-title mb-4 text-center"
              >
                Login
              </h2>
              <form>
                <div
                  class="mb-3"
                >
                  <label
                    class="form-label"
                    for="login-email"
                  >
                    Email:
                  </label>
                  <input
                    class="form-control"
                    id="login-email"
                    placeholder="Enter your email"
                    required=""
                    type="email"
                    value=""
                  />
                </div>
                <div
                  class="mb-3"
                >
                  <label
                    class="form-label"
                    for="login-password"
                  >
                    Password:
                  </label>
                  <input
                    class="form-control"
                    id="login-password"
                    placeholder="Enter your password"
                    required=""
                    type="password"
                    value=""
                  />
                </div>
                <button
                  class="btn btn-primary w-100"
                  type="submit"
                >
                  Login
                </button>
              </form>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
